File: ch5/python/code5_3.py
# 예제 코드를 일부 수정합니다. 책에는 0으로 되어 있습니다만
# 초기 위치와 도착 위치가 동일한 경우 경로의 수는 가만히 있는 1 입니다.
# 오류를 사과드립니다.
# 책의 예제에는 메모이제이션 예제가 아닙니다만, 메모이제이션 예제로 수정해서 보여 드립니다.

# 메모이제이션 + 하향식 코드
def num_of_paths(m, n, memo = None):
    if memo is None:
        memo = {}
    if (m == 0) or (n == 0):
        return 1
    if (m, n) in memo:
        return memo[(m, n)]
    memo[(m, n)] = num_of_paths(m - 1, n, memo) + \
                   num_of_paths(m, n - 1, memo)
    return memo[(m, n)]

M, N = input('방의 구조를 입력하세요 : ').split()
M = int(M)
N = int(N)
print('총 경로의 수는 %d개입니다.' % num_of_paths(M - 1, N - 1))

File: ch5/c/code5_3.c
#include <stdio.h>

int numOfPaths(int m, int n){
  // 종료 조건
  if(m == 0 && n == 0)  // 방 (0, 0)
    return 0;
  if(m == 0 || n == 0)  // 첫 번째 행 또는 첫 번째 열
    return 1;

  // 재귀 호출
  return numOfPaths(m - 1, n) + numOfPaths(m, n - 1);
}

int main(){
  int M, N;
  printf("방의 구조를 입력하세요 : ");
  scanf("%d %d", &M, &N);
  // 인덱스가 0 부터 이므로 M - 1, N - 1을 파라미터로 호출합니다.
  printf("총 경로의 수는 %d개 입니다.\n", numOfPaths(M - 1, N - 1));
  return 0;
}